src/finalizer/index.ts

```diff
@@ -31,18 +33,121 @@ import {
   FINALIZER_TOKENBRIDGE_LOOKBACK,
   Multicall2Call,
 } from "../common";
+import { ChainFinalizer, FinalizerPromise, Withdrawal } from "./types";
+
 config();
 let logger: winston.Logger;
 
-export interface Withdrawal {
-  l2ChainId: number;
-  l1TokenSymbol: string;
-  amount: string;
-}
-
 // Filter for optimistic rollups
 const oneDaySeconds = 24 * 60 * 60;
 
+async function optimismFinalizer(
+  signer: Wallet,
+  hubPoolClient: HubPoolClient,
+  spokePoolClient: SpokePoolClient,
+  latestBlockToFinalize: number
+): Promise<FinalizerPromise> {
+  const { chainId } = spokePoolClient;
+  assert(chainId === 10, `Chain ID mismatch: ${chainId} != 10`);
+
+  const crossChainMessenger = getOptimismClient(10, signer);
+
+  // Sort tokensBridged events by their age. Submit proofs for recent events, and withdrawals for older events.
+  const earliestBlockToProve = latestBlockToFinalize + 1;
+  const { recentTokensBridgedEvents = [], olderTokensBridgedEvents = [] } = groupBy(
+    spokePoolClient.getTokensBridged(),
+    (e) => (e.blockNumber >= earliestBlockToProve ? "recentTokensBridgedEvents" : "olderTokensBridgedEvents")
+  );
+
+  // First submit proofs for any newly withdrawn tokens. You can submit proofs for any withdrawals that have been
+  // snapshotted on L1, so it takes roughly 1 hour from the withdrawal time
+  logger.debug({
+    at: "Finalizer",
```

Comment:
Do we want to specify the function we're in?
```suggestion
    at: "Finalizer#optimismFinalizer",
```

Reply:
yes please

Reply:
Implemented: https://github.com/across-protocol/relayer-v2/pull/851/commits/a1887cc9c2bbfb7b0a650af8641f5e80bd177fc3